homeassistant/components/esphome/bluetooth.py

```diff
@@ -1,59 +1,104 @@
 """Bluetooth scanner for esphome."""
+from __future__ import annotations
 
 from collections.abc import Callable
 import datetime
 from datetime import timedelta
 import re
 import time
+from typing import Any
+import uuid
 
 from aioesphomeapi import APIClient, BluetoothLEAdvertisement
+from bleak.backends.characteristic import BleakGATTCharacteristic
+from bleak.backends.client import BaseBleakClient
 from bleak.backends.device import BLEDevice
 from bleak.backends.scanner import AdvertisementData
+from bleak.backends.service import BleakGATTServiceCollection
 
 from homeassistant.components.bluetooth import (
     BaseHaScanner,
+    HaBluetoothConnector,
     async_get_advertisement_callback,
     async_register_scanner,
 )
 from homeassistant.components.bluetooth.models import BluetoothServiceInfoBleak
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.core import CALLBACK_TYPE, HomeAssistant, callback
+from homeassistant.core import (
+    CALLBACK_TYPE,
+    HomeAssistant,
+    async_get_hass,
+    callback as hass_callback,
+)
 from homeassistant.helpers.event import async_track_time_interval
 
+from .domain_data import DomainData
+from .entry_data import RuntimeEntryData
+
 ADV_STALE_TIME = 180  # seconds
 
 TWO_CHAR = re.compile("..")
 
 
+@hass_callback
+def async_can_connect(source: str) -> bool:
+    """Check if a given source can make another connection."""
+    domain_data = DomainData.get(async_get_hass())
+    client = domain_data.get_entry_data(domain_data.get_by_unique_id(source)).client
+    return bool(client.available_ble_connections)
```

Comment:
This also needs to handle the case where the esphome device is no longer available because it's lost power or network